src/com/google/enterprise/gsafeed/GsafeedHelper.java

```diff
@@ -0,0 +1,125 @@
+package com.google.enterprise.gsafeed;
+
+import org.w3c.dom.Document;
+import org.xml.sax.EntityResolver;
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.StringReader;
+import java.net.URL;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+/**
+ * Helper methods for reading and creating GSA feed files.
+ */
+public class GsafeedHelper {
+
+  public static Gsafeed unmarshalWithDtd(URL url) throws JAXBException,
+      IOException, ParserConfigurationException, SAXException {
+    return unmarshalWithDtd(url.openStream());
+  }
+
+  /**
+   * Use the DTD to check for errors in the feed being read.
+   * Assume the DTD is on the classpath as /gsafeed.dtd.
+   */
+  public static Gsafeed unmarshalWithDtd(InputStream inputStream)
+      throws JAXBException, IOException, ParserConfigurationException,
+      SAXException {
+    JAXBContext jaxbContext =
+        JAXBContext.newInstance("com.google.enterprise.gsafeed");
+    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
+    DocumentBuilderFactory docBuilderFactory =
+        DocumentBuilderFactory.newInstance();
+    docBuilderFactory.setValidating(true);
+    DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
+    docBuilder.setEntityResolver(new EntityResolver() {
+        @Override
+        public InputSource resolveEntity(String publicId,
+            String systemId) throws SAXException, IOException {
+          if (publicId.equals("-//Google//DTD GSA Feeds//EN")) {
+            return new InputSource(
+                GsafeedHelper.class.getResource("/gsafeed.dtd").openStream());
+          }
+          return null; // DTD is the only expected entity in feed files
+        }
+      });
+    docBuilder.setErrorHandler(new ErrorHandler() {
+        @Override
+        public void warning(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+
+        @Override
+        public void error(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+
+        @Override
+        public void fatalError(SAXParseException exception)
+            throws SAXException {
+          throw new SAXException(exception.toString());
+        }
+      });
+    Document document = docBuilder.parse(inputStream);
+    return (Gsafeed) unmarshaller.unmarshal(document);
+  }
+
+  /**
+   * Avoid reading the DTD. No validation will happen.
```

Comment:
When do we want to use the DTD and when do we not want to use it?

Reply:
You could potentially read a less-than-perfect feed file if you bypass validation, and then fix any issues in code. I don't know if that's a realistic use case. You might not bother with validation when reading files from the GSA.